### ISSUE TITLE
Allow early access to nghttp2_frame_hd

### ISSUE THREAD

**Issue by chen-wang** [nghttp2/nghttp2#72]
Is it possible to add a new callback API "nghttp2_on_begin_frame_callback" to "nghttp2_session_callbacks", which is called whenever a frame header is received?

This allows users to access the frame header as soon as it becomes available and be able to
1. learn the frame length and pre-allocate a data buffer in a single memory allocation to receive data chunks received in nghttp2_on_data_chunk_recv_callback.
2. learn the frame type and flags before receiving payload, which may be useful in HTTP/2 extension.

Basically, this API is an extension of "nghttp2_on_begin_headers_callback" to allow users to access frame headers for every frames instead of just header frames.

Thanks
Chen
